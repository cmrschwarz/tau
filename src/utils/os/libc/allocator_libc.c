#include "../../allocator.h"
#ifdef USE_LIBC_ALLOCATOR
 
#include <memory.h> 
#include "../../math_utils.h"

//this is a really basic portable allocator using libc
//this is used as a fallback if no os specific one has been written yet
//to get rid of the libc dependency

static struct allocator{
    ureg segment_size;
}ALLOCATOR;

int allocator_init(){
    ALLOCATOR.segment_size = 4096;
    return 0;
}

//TODO(cmrs): debugging feature counting allocations
//that complains here if we had fewer or more deallocations --> leaks
void allocator_fin(){}

ureg allocator_get_segment_size(){
    return ALLOCATOR.segment_size;
}

int tal_init(thread_allocator* tal){
    return 0;
}

void tal_fin(thread_allocator* tal){}


int tal_alloc(thread_allocator* tal, ureg size, memblock* b){
    b->start = malloc(size);
    if(b->start == NULL)return -1;
    b->end = ptradd(b->start, size);
    return 0;
}

int tal_realloc(thread_allocator* tal, ureg used_size, ureg new_size, memblock* b) {
    void* new = realloc(b->start, new_size);
    if(!new)return -1;
    b->start = new;
    b->end = ptradd(new, new_size);
    return 0;
}

void tal_free(thread_allocator* tal, memblock* b){
    free(b->start);
}

#endif