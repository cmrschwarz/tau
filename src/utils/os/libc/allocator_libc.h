#pragma once
#define USE_LIBC_ALLOCATOR

typedef struct thread_allocator {
    // this doesn't need anything for now
} thread_allocator;
