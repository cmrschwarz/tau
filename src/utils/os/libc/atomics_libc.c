#include "../../threading.h"
#ifdef USE_LIBC_ATOMICS
// ptr
int atomic_ptr_init(atomic_ptr* a, void* value)
{
    atomic_init(a, value);
    return 0;
}
void* atomic_ptr_load(atomic_ptr* a)
{
    return atomic_load(a);
}
void atomic_ptr_store(atomic_ptr* a, void* value)
{
    atomic_store(a, value);
}
bool atomic_ptr_cas(atomic_ptr* a, void** oldval, void* newval)
{
    return atomic_compare_exchange_weak(a, oldval, newval);
}
void atomic_ptr_fin(atomic_ptr* a)
{
}

// bool
int atomic_bool_init(atomic_bool* a, bool value)
{
    atomic_init(a, value);
    return 0;
}
bool atomic_bool_load(atomic_bool* a)
{
    return atomic_load(a);
}
void atomic_bool_store(atomic_bool* a, bool value)
{
    atomic_store(a, value);
}
bool atomic_bool_cas(atomic_bool* a, bool* oldval, bool newval)
{
    return atomic_compare_exchange_weak(a, oldval, newval);
}
void atomic_bool_fin(atomic_bool* a)
{
}

// ureg
int atomic_ureg_init(atomic_ureg* a, ureg value)
{
    atomic_init(a, value);
    return 0;
}
ureg atomic_ureg_load(atomic_ureg* a)
{
    return atomic_load(a);
}
void atomic_ureg_store(atomic_ureg* a, ureg value)
{
    atomic_store(a, value);
}
ureg atomic_ureg_inc(atomic_ureg* a)
{
    return atomic_fetch_add(a, 1);
}
ureg atomic_ureg_dec(atomic_ureg* a)
{
    return atomic_fetch_sub(a, 1);
}
ureg atomic_ureg_add(atomic_ureg* a, ureg v)
{
    return atomic_fetch_add(a, v);
}
ureg atomic_ureg_sub(atomic_ureg* a, ureg v)
{
    return atomic_fetch_sub(a, v);
}
bool atomic_ureg_cas(atomic_ureg* a, ureg* oldval, ureg newval)
{
    return atomic_compare_exchange_weak(a, oldval, newval);
}
void atomic_ureg_fin(atomic_ureg* a)
{
}

// sreg
int atomic_sreg_init(atomic_sreg* a, sreg value)
{
    atomic_init(a, value);
    return 0;
}
sreg atomic_sreg_load(atomic_sreg* a)
{
    return atomic_load(a);
}
void atomic_sreg_store(atomic_sreg* a, sreg value)
{
    atomic_store(a, value);
}
sreg atomic_sreg_inc(atomic_sreg* a)
{
    return atomic_fetch_add(a, 1);
}
sreg atomic_sreg_dec(atomic_sreg* a)
{
    return atomic_fetch_sub(a, 1);
}
sreg atomic_sreg_add(atomic_sreg* a, sreg v)
{
    return atomic_fetch_add(a, v);
}
sreg atomic_sreg_sub(atomic_sreg* a, sreg v)
{
    return atomic_fetch_sub(a, v);
}
bool atomic_sreg_cas(atomic_sreg* a, sreg* oldval, sreg newval)
{
    return atomic_compare_exchange_weak(a, oldval, newval);
}
void atomic_sreg_fin(atomic_sreg* a)
{
}

#endif