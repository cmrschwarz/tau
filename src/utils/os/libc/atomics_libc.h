#pragma once

#ifndef __STDC_NO_ATOMICS__
#include <stdatomic.h>
#endif

#define USE_LIBC_ATOMICS

typedef _Atomic ureg atomic_ureg;
typedef _Atomic sreg atomic_sreg;
typedef _Atomic bool atomic_bool;