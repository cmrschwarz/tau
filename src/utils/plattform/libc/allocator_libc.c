#include "../../allocator.h"
#ifdef USE_LIBC_ALLOCATOR

#include "../../math_utils.h"

#include <memory.h>
#if DEBUG
#define PRINT_ALLOCS 0
#define DISABLE_LEAKCHECK 0
#include "../../panic.h"
#include "../../threading.h"
#include "utils/debug_utils.h"
#include <stdio.h>
static atomic_sreg allocations;
#endif

#include "../../../error_log.h"
static master_error_log* mel = NULL;

static inline void count_alloc(void* v, ureg size)
{
#if DEBUG
    atomic_sreg_inc(&allocations);
#if PRINT_ALLOCS
    tprintf("alloc: %zx [%zu B]\n", v, size);
#endif
#endif
}
static inline void count_free(void* v)
{
#if DEBUG
    atomic_sreg_dec(&allocations);
#if PRINT_ALLOCS
    tprintf("free: %zx\n", v);
#endif
#endif
}
bool talloc_initialized()
{
    return mel != NULL;
}
int talloc_init(master_error_log* el)
{
#if DEBUG
    atomic_sreg_init(&allocations, 0);
#endif
    mel = el;
    return OK;
}

void talloc_fin()
{
    assert(mel);
    mel = NULL;
#if DEBUG && !DISABLE_LEAKCHECK
    sreg a = atomic_sreg_load(&allocations);
    if (a) {
        fprintf(stderr, "MEMORY LEAK! (allocs - frees) = %zd\n", a);
        panic("Memory leak!");
    }
#endif
}

void* tmalloc(ureg size)
{
    void* r = malloc(size);
    if (r) {
        count_alloc(r, size);
    }
    else {
        assert(mel);
        aseglist_iterator it;
        aseglist_iterator_begin(&it, &mel->error_logs);
        ureg tid = thread_id();
        for (error_log* el = aseglist_iterator_next(&it); el != NULL;
             el = aseglist_iterator_next(&it)) {
            if (el->tid == tid) {
                error_log_report_allocation_failiure(el);
                return NULL;
            }
        }
        master_error_log_report(mel, "allocation failure");
        return NULL;
    }
    return r;
}

void* tmallocz(ureg size)
{
    void* m = tmalloc(size);
    if (m != NULL) memset(m, 0, size);
    return m;
}

void* trealloc(void* old, ureg used_size, ureg new_size)
{
    return realloc(old, new_size);
}

void tfree(void* mem)
{
    assert(mem != NULL);
    count_free(mem);
    free(mem);
}

#endif
