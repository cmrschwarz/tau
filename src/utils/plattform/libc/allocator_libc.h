#pragma once
#define USE_LIBC_ALLOCATOR