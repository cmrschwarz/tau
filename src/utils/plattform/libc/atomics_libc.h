#pragma once
#include "../../plattform.h"
#include "../../types.h"

#define USE_LIBC_ATOMICS

#if CMPLR_CLANG
#   define ATOMIC_PREFIX _Atomic
#elif CMPLR_GCC 
#   define ATOMIC_PREFIX volatile
#else
#   error tauc has no libc atomics implementation on this plattform
#endif

typedef struct atomic_ureg_s {
    ATOMIC_PREFIX ureg val;
} atomic_ureg;
typedef struct atomic_sreg_s {
    ATOMIC_PREFIX sreg val;
} atomic_sreg;
typedef struct atomic_ptr_s {
    void* ATOMIC_PREFIX val;
} atomic_ptr;
// we can't name it atomic_bool since that's used in stdatomic and would cause
// all kinds of problems...
typedef struct atomic_boolean_s {
    ATOMIC_PREFIX bool val;
} atomic_boolean;
