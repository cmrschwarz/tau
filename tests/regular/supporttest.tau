require "../support/libc.tau";
import libc::(printf, puts, malloc, free, debug_silence_stdout);

public func foo() -> int{
    printf("printf with no args\n");
    printf("printf arg1: %i\n", 1);
    printf("printf arg1: %i arg2: %i\n", 1, 2);
    puts("puts");
    x := malloc(16) as *[2]int;
    printf("malloced array\n");
    (*x)[0] = 42;
    (*x)[1] = 17;
    printf("filled mallloced array: {%i, %i}\n", (*x)[0], (*x)[1]);
    res := (*x)[0] - 42 + (*x)[1] - 17;
    free(x as *void);
    return res;
}

public extern func main() -> int{
    foo();
    #debug_silence_stdout();
    return #foo();
}
