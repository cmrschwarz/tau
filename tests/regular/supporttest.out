printf with no args
printf arg1: 1
printf arg1: 1 arg2: 2
puts
malloced array
filled mallloced array: {42, 17}
